searchlib/src/tests/queryeval/nearest_neighbor/nearest_neighbor_test.cpp

```diff
@@ -0,0 +1,177 @@
+
+    Config _cfg;
+    vespalib::string _name;
+    vespalib::string _typeSpec;
+    std::shared_ptr<DenseTensorAttribute> _tensorAttr;
+    std::shared_ptr<AttributeVector> _attr;
+
+    Fixture(const vespalib::string &typeSpec)
+        : _cfg(BasicType::TENSOR, CollectionType::SINGLE),
+          _name("test"),
+          _typeSpec(typeSpec),
+          _tensorAttr(),
+          _attr()
+    {
+        _cfg.setTensorType(ValueType::from_spec(typeSpec));
+        _tensorAttr = makeAttr();
+        _attr = _tensorAttr;
+        _attr->addReservedDoc();
+    }
+
+    ~Fixture() {}
+
+    std::shared_ptr<DenseTensorAttribute> makeAttr() {
+        return std::make_shared<DenseTensorAttribute>(_name, _cfg);
+    }
+
+    void ensureSpace(uint32_t docId) {
+        while (_attr->getNumDocs() <= docId) {
+            uint32_t newDocId = 0u;
+            _attr->addDoc(newDocId);
+            _attr->commit();
+        }
+    }
+
+    void setTensor(uint32_t docId, const Tensor &tensor) {
+        ensureSpace(docId);
+        _tensorAttr->setTensor(docId, tensor);
+        _attr->commit();
+    }
+
+    void setTensor(uint32_t docId, double v1, double v2) {
+        auto t = createTensor(TensorSpec(denseSpec)
+                              .add({{"x", 0}}, v1)
+                              .add({{"x", 1}}, v2));
+        setTensor(docId, *t);
+    }
+};
+
+template <bool strict>
+SimpleResult find_matches(Fixture &env) {
+    using NNI = NearestNeighborIterator<strict>;
+    auto md = MatchData::makeTestInstance(2, 2);
+    auto qt = createTensor(TensorSpec(denseSpec));
+
+    auto &tfmd = *(md->resolveTermField(0));
+    const DenseTensorView &qtv = *qt;
+    auto &attr = *(env._tensorAttr);
+
+    NearestNeighborDistanceHeap dh(2);
+    NNI search(tfmd, qtv, attr, dh);
+    if (strict) {
+        return SimpleResult().searchStrict(search, attr.getNumDocs());
+    } else {
+        return SimpleResult().search(search, attr.getNumDocs());
+    }
+}
+
+TEST("require that NearestNeighborIterator returns expected results") {
+    Fixture fixture(denseSpec);
+    fixture.ensureSpace(6);
+    fixture.setTensor(1, 3.0, 4.0);
+    fixture.setTensor(2, 6.0, 8.0);
+    fixture.setTensor(3, 5.0, 12.0);
+    fixture.setTensor(4, 4.0, 3.0);
+    fixture.setTensor(5, 8.0, 6.0);
+    fixture.setTensor(6, 4.0, 3.0);
+    SimpleResult result = find_matches<true>(fixture);
+    SimpleResult expect({1,2,4,6});
+    EXPECT_EQUAL(result, expect);
+    result = find_matches<false>(fixture);
+    EXPECT_EQUAL(result, expect);
+}
+
+template <bool strict>
+std::vector<feature_t> get_rawscores(Fixture &env) {
+    using NNI = NearestNeighborIterator<strict>;
+    auto md = MatchData::makeTestInstance(2, 2);
+    auto qt = createTensor(TensorSpec(denseSpec));
+    auto &tfmd = *(md->resolveTermField(0));
+    const DenseTensorView &qtv = *qt;
+    auto &attr = *(env._tensorAttr);
+    NearestNeighborDistanceHeap dh(2);
+    NNI search(tfmd, qtv, attr, dh);
```

Comment:
This setup is equal to the one in find_matches(). Consider refactor into a class/struct to avoid code duplication.

Reply:
i tried this and the result was a lot more code that was much more unreadable.